File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> nodes;
        Node *temp = head;
        while(temp){
            nodes[temp] = new Node(temp->val);
            temp = temp->next;
        }

        temp = head;
        while(temp){
            Node *new_node = nodes[temp];
            new_node->next = nodes[temp->next];
            new_node->random = nodes[temp->random];
            temp = temp->next;
        }

        return nodes[head];
    }
};

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return head
    }
    mpp := make(map[*Node]*Node)

    newHead := &Node{head.Val, nil, nil}
    mpp[head] = newHead

    dummy := head

    for dummy != nil {
        if _, ok := mpp[dummy]; !ok {
            newNode := &Node{dummy.Val, nil, nil}

            mpp[dummy] = newNode
        }

        dummy = dummy.Next
    }

    for key, val := range mpp {
        node := val

        node.Next = mpp[key.Next]
        node.Random = mpp[key.Random]
    }

    return newHead
}